pyccl/halos/pk_4pt.py

```diff
@@ -505,3 +585,976 @@ def _logged_output(*arrs, log):
+            a profile covariance object returning the the two-point
+            moment of `prof` and `prof2`. If `None`, the default
+            second moment will be used, corresponding to the
+            products of the means of both profiles.
+        prof3 (:class:`~pyccl.halos.profiles.HaloProfile`): halo
+            profile (corresponding to :math:`v_1` above. If `None`,
+            `prof` will be used as `prof3`.
+        prof4 (:class:`~pyccl.halos.profiles.HaloProfile`): halo
+            profile (corresponding to :math:`v_2` above. If `None`,
+            `prof3` will be used as `prof4`.
+        prof13_2pt (:class:`~pyccl.halos.profiles_2pt.Profile2pt`):
+            same as `prof12_2pt` for `prof` and `prof3`.
+        prof14_2pt (:class:`~pyccl.halos.profiles_2pt.Profile2pt`):
+            same as `prof14_2pt` for `prof` and `prof4`.
+        prof24_2pt (:class:`~pyccl.halos.profiles_2pt.Profile2pt`):
+            same as `prof14_2pt` for `prof2` and `prof4`.
+        prof32_2pt (:class:`~pyccl.halos.profiles_2pt.Profile2pt`):
+            same as `prof14_2pt` for `prof3` and `prof2`.
+        p_of_k_a (:class:`~pyccl.pk2d.Pk2D`): a `Pk2D` object to
+            be used as the linear matter power spectrum. If `None`, the power
+            spectrum stored within `cosmo` will be used.
+
+    Returns:
+        float or array_like: integral values evaluated at each
+        combination of `k` and `a`. The shape of the output will
+        be `(N_a, N_k, N_k)` where `N_k` and `N_a` are the sizes of
+        `k` and `a` respectively. The ordering is such that
+        `output[ia, ik2, ik1] = T(k[ik1], k[ik2], a[ia])`
+        If `k` or `a` are scalars, the corresponding dimension will
+        be squeezed out on output.
+    """
+    a_use = np.atleast_1d(a)
+    k_use = np.atleast_1d(k)
+
+    # Check inputs
+    prof, prof2, prof3, prof4, prof13_2pt, prof14_2pt, \
+        prof24_2pt, prof32_2pt = _allocate_profiles2(prof, prof2, prof3, prof4,
+                                                     prof13_2pt, prof14_2pt,
+                                                     prof24_2pt, prof32_2pt)
+
+    na = len(a_use)
+    nk = len(k_use)
+
+    # Power spectrum
+    def get_isotropized_pkr(p_of_k_a, aa):
+        def integ(theta):
+            mu = np.cos(theta)
+            k = np.sqrt(k_use[:, None]**2+k_use[None, :]**2
+                        + 2*k_use[None, :]*k_use[:, None]*mu)
+            kk = k.flatten()
+            if isinstance(p_of_k_a, Pk2D):
+                pk = p_of_k_a(kk, aa, cosmo)
+            elif (p_of_k_a is None) or (str(p_of_k_a) == 'linear'):
+                pk = cosmo.get_linear_power()(kk, aa)
+            elif str(p_of_k_a) == 'nonlinear':
+                pk = cosmo.get_nonlin_power()(kk, aa)
+            else:
+                raise TypeError("p_of_k_a must be `None`, \'linear\', "
+                                "\'nonlinear\' or a `Pk2D` object")
```

Comment:
could we set up the Pk2D outside of the integrand (so it only happens once)? Should help  with speed and would be a bit neater.

Reply:
Done